### PR TITLE
Scale and Mask

### PR DESCRIPTION
**Pull request by LeicongLi**
## What changes were proposed in this pull request?
1. added 5 fields into protobuf definition, inputDimMasks, inputScales, outputDimMasks, outputScales, isMklInt8Enabled; updated related auto-generated code.
2. added 1 Trait MklInt8Convertible which hold data and methods to calculate required information for converting from fp32 to int8.
3. added functions to serialize and deserialize MKL-DNN Int8 related values 
4. made MKL-DNN Linear module inherit MKL-DNN Int8 Trait
(Please fill in changes proposed in this patch)

## How was this patch tested?
1. Construct a module, Linear in my case.
2. Generate and hold random values
2. Assign random values to above fields
3. Save the module
4. Load the module
5. Verify the values stored in the fields are the same as assigned at step #2.

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If it is possible, please attach a screenshot; otherwise, remove this)

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -907,6 +908,20 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
 
   // ================================= Internal APIs ===========================================
 
+  private var protobufTest : String = Integer.toHexString(java.util.UUID.randomUUID().hashCode())
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: what is this for ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala [hunk 2]
```diff
@@ -356,12 +383,44 @@ trait ModuleSerializable extends Loadable with Savable{
...
+    * and load them into BigDL Module object
+    * @param context deserialized context
+    * @param module  bigDL Module with relationships
+    */
+  private def loadScaleAndMask[T: ClassTag](context: DeserializeContext,
+                                            module: AbstractModule[Activity, Activity, T])
+                                           (implicit ev: TensorNumeric[T]): Unit = {
+
+    val protobufModel = context.bigdlModule
+    module.setProtobufTest(protobufModel.getProtobufTest)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala]: not needed ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 3]
```diff
@@ -1181,5 +1196,26 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
     }
   }
+
+
+  // this three variables are for mkldnn int8. because container maybe have more than 1 Tensor
+  // output, so use a mutable ArrayBuffer[Float] to handle the ConcatTable, Graph and so on
+  // mask == 0 means we use single scales, which will get the max of whole tensor (in/out).
+  // scalesOfInput will maintain the max value of input tensor
+  // scalesOfOutput will maintain the max value of output tensor
+  var scalesOfInput: Int8ScalesAndMask = new Int8ScalesAndMask(Int8ScalesAndMask.SINGLE_SCALE)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: I prefer not initializing it here as it's only used in Int8, they are null for other cases

**Comment by wzhongyuan** [comment]
Please add unit tests for it

**Comment by jason-dai** [comment]
I don't think we should put these in `AbstractModule`; instead, define a subclass like `MklInt8Module`, similar to `QuantizedModule` or `MklDnnLayer`

**Comment by wzhongyuan** [comment]
@jason-dai    yes, that sounds better

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MklInt8Convertible.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+  /**
+   * Scales update helper. Replace scale at specific index with provided new scale
+   * @param scales the scales arrayBuffer to be updated
+   * @param scale the new scale
+   * @param index the index of which the scale need to be updated
+   * @return Unit
+   */
+  private def updateScalesHelper(scales: ArrayBuffer[Array[Float]],
+                         scale: Array[Float], index: Int): Unit = {
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MklInt8Convertible.scala]: The indention of scale is not right.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala [hunk 5]
```diff
@@ -232,41 +238,63 @@ trait ModuleSerializable extends Loadable with Savable{
...
+
+  /*
+   * Create BigDL model's protobuf definition by serializing BigDL Module object
+   * @param BigDLModule.Builder: BigDL model builder of protobuf definition
+   * @param SerializeContext: Serialized context of BigDL module
+   * @return SerializeResult:
+   */
+  protected def createSerializeBigDLModule[T: ClassTag](modelBuilder : BigDLModule.Builder,
+                                      context: SerializeContext[T])
+                                    (implicit ev: TensorNumeric[T]): SerializeResult = {
```
Comment (i8run) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala]: The syntax is a little strange.

Comment (LeicongLi) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/serializer/ModuleSerializable.scala]: Thanks. I will modify these syntax issues.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MklInt8Convertible.scala [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+package com.intel.analytics.bigdl.nn.mkldnn
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+* Trait which provides MKL-DNN functionality to convert FP32 model to INT8 model
+*/
+trait MklInt8Convertible {
+  // input dimension mask
+  protected var inDimMask: Int = 0
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MklInt8Convertible.scala]: By BigDL conversion, I prefer to use `inputDimMask`, similar below for `outputDimMask`

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/mkldnn/Fp32ToInt8Spec.scala [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import scala.util.Random
+
+class Fp32ToInt8Spec  extends FlatSpec with Matchers with BeforeAndAfter {
+
+  val modelPath: String = "myTestModel" + UUID.randomUUID().toString
+  val weightPath: String = "myTestModelWeight" + UUID.randomUUID().toString
+
+  "Saving and loading scale and mask" should "work properly" in {
+
+    val myModel = Linear(3, 4)
```
Comment (wzhongyuan) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/mkldnn/Fp32ToInt8Spec.scala]: add a unit test for empty scales

**Comment by wzhongyuan** [comment]
I think there are other layers that need to implement this new trait. Also for mkl-blas layers, we should add that. @i8run
